regression function

changsiyao opened the pull request:
Here are the regression functions I wrote.
Need to use Ben's new event2neural function and merge_cond function.
These works on my computer, but somebody should check if things make sense. Otherwise, I'll look more carefully at these codes over the weekend.

changsiyao committed:
changed boundary for outliers

changsiyao committed:
imported function from graph_function file

changsiyao committed:
add function to find outliers based on dvars and fd measurements

changsiyao committed:
add script to implement outlier functions

changsiyao committed:
update boundary option

changsiyao committed:
add ing a boundary option for the function

changsiyao committed:
update boundary option

changsiyao committed:
Merge branch 'master' of https://github.com/berkeley-stat159/project-theta

changsiyao committed:
Merge branch 'dev' of https://github.com/berkeley-stat159/project-theta into placeholder

changsiyao committed:
remove code handled by Ben

changsiyao committed:
remove code handled by Ben

changsiyao committed:
Merge branch 'dev' of https://github.com/berkeley-stat159/project-theta into placeholder

changsiyao committed:
add regression functions and scripts

changsiyao committed:
add regression functions and scripts

BenjaminHsieh commented:
Ok, should we see if #34 can be merged first? b/c of eventstoneural_extended
Also:
Question: Do you think we can figure out how to plot multiple plots in on figure? This way we dont need to output/save so many files, such as the the massive for loops we have in plotting meanSig, fd, dvars; will probably try to implement this sometime tomorrow if people agree

BenjaminHsieh commented on code/outlierfunction.py:
```diff
@@ -0,0 +1,19 @@
+            outlier.append(i)
+    # find outlier indices
+    outlier_indices = np.nonzero(outlier)
+    # write file to store outlier indices
```
Do we these three lines below inside the function? do you think it may be better to just output indicies and actually save in the script (create an array that stores the list of indicies from each run and save this array as txt), or include a boolean indicator in parameter whether to save or not.

changsiyao replied:
I can try fixing it tmr, probably save everything in a dictionary, like {'sub1run1': outliers, 'sub1run2': outliers, ...}. Is this format ok?

BenjaminHsieh replied:
yeah something like that, whatever is more convenient to access if we are to do plotting/removal of outliers. I am not sure if a dictionary save/load properly from txt, try it out

changsiyao commented:
As for figures, I guess we can add a boolean on whether save of not (like I think you already did), cuz figures are just helpful for visualization. We don't necessarily need the figures to identify outliers.
But by plotting them together, do you mean if we can have 3 subplots side by side in 1 figure? I think it's do-able, we just need to figure out how. It's probably something like, fig = plt.figure(); ax = fig.add_subplot(131); fig.add_subplot(132); fig.add_subplot(133)... we would need to google around for that.

BenjaminHsieh commented:
The reason I dont really like saving things inside functions is because it is hard to write test function for them. But yeah boolean would be good, if false we can just return the values instead. 
Yes, 3 subplots side by side works, or can do even more and create a lattice of subplots, yeah have to look into how to do that but I think I might be able to steal some ideas off of other groups, lol

changsiyao committed:
Merge branch 'dev' of https://github.com/berkeley-stat159/project-theta into placeholder

changsiyao committed:
Merge branch 'dev' of https://github.com/berkeley-stat159/project-theta into placeholder

briankleinqiu commented:
Awesome work! Just out of curiosity, is there an advantage to doing it this way and not using something like this?

http://statsmodels.sourceforge.net/devel/regression.html

changsiyao committed:
update outlier function and script to put in a dictionary

changsiyao committed:
update graph functions to label outliers

changsiyao commented:
1. Updated the outliers functions, so now they should be stored in a txt file like a python dictionary.
2. Updated the graph functions to label outliers.
3. To Brian's suggestion: I was doing the regression like the hw, so I just didn't search around for packages. But I guess we could use already existed packages for regressions. Let me google around and try it.

BenjaminHsieh commented:
If both methods work, we can use one way as a test function for the other way

BenjaminHsieh commented:
i can write the test functions for regression scripts if you havent done so already

changsiyao commented:
That would be great, cuz I haven't done it yet.

briankleinqiu commented:
Minor error, the regression_script.py calls in behavtask_tr.py, but that's in utils while the script is in code and so there's a path error for me

BenjaminHsieh commented on code/outlierfunction.py:
```diff
@@ -0,0 +1,16 @@
+
+# find outliers based on DVARS and FD
+
+def outlier(file_name, bound):
```
Can we make change outlier take in arrays instead of files? load the txt file in the script. This way we dont have to create a text file every time we run the test. In general, try to make function parameters be data types, not files;  thanks

changsiyao committed:
moved outlier and regression functions and scripts to utils folder

changsiyao committed:
update function and comment

changsiyao committed:
fixed path

changsiyao commented:
moved all the codes to the utils folder and changed some paths in the codes accordingly. Hopefully, all the path are correct now.

briankleinqiu commented:
Here are some more minor errors I found: 

## REGRESSION_FUNCTIONS.PY

-regression_functions.py needs to import events2neural_extend and merge_cond at the top

-in getGainLoss function, hrf_at_times is undefined. Did you mean to set hrf_at_times = hrf(tr_times) in the script? I see you define hrf_at_times in the regression script, but it won't recognize since hrf_times in getGainLoss is local and not global. I recommend we define hrf_times within the function itself or something similar

-in getGainLoss all_tr_times is defined but never used. all_tr_times is again defined in a different scope in the script. Let's pick one and stick with that

-in calcBeta and calcMRSS, data is again undefined. We should define data within the function, because right now we're defining data outside of that scope and doing it in the script, but i don't think it will run correctly because of that. 

In general perhaps it would be better practice if we made data an argument to the functions and passed it in through there. Same with all_tr_times

## REGRESSION_SCRIPT.PY

behavtask_tr.py should just be behavtask_tr i believe

I can clean this up and make a new pull request if you're pressed for time, but I wanted to check in and see if there was a reason for defining data and all_tr_times outside of the functions before making changes. Also I think Ben's thoughts on changing the parameter from the file to actual arrays is good

changsiyao commented:
Sorry about those, let me take a look and fix these.

changsiyao commented:
Actually, do you have the problem that if you don't specifically import numpy as np inside a function, the function won't find np?

changsiyao committed:
fix errors

briankleinqiu commented:
hmmm no i don't have that problem. what's the exact wording of the error? import numpy as np is at the top of everything so it should be fine right

changsiyao committed:
imported numpy

changsiyao commented:
I fixed these functions again. Can someone check these functions?

BenjaminHsieh commented on code/utils/regression_functions.py:
```diff
@@ -0,0 +1,72 @@
+    designp = npl.pinv(design)
+    T = data.shape[-1]
+    time_by_vox = np.reshape(data, (-1, T)).T
+    beta_hats = designp.dot(time_by_vox)
```
Also are lines 60 - 66 just the same as the return from calcBeta? can we just call calcBeta?

changsiyao committed:
Merge branch 'dev' of https://github.com/berkeley-stat159/project-theta into placeholder

BenjaminHsieh commented:
Let me know when you get a chance to update these, I have some test functions written that we can check.

changsiyao committed:
fixed X design and some functions

changsiyao committed:
add outlier dictionary

changsiyao commented:
Updated the function with X design being the task condition data. 
Also, switched the events2neural and merge_cond function into the script rather than in another function.
I checked a couple diagnostic stats, still pretty bad. Jiang is working on the inferences testing. We'll get some result tmr or Friday.

changsiyao commented:
Also, here are some code if we try to use sklearn or statsmodels.api package.
Didn't want to pull request on this.

from sklearn import datasets, linear_model

// X, Y are constructed like the way in calcBeta(data, gain, loss).
X = np.ones((len(gain), 3))
X[:, 0] = gain
X[:, 1] = loss
T = data.shape[-1]
Y = np.reshape(data, (-1, T)).T

// Create linear regression object
regr = linear_model.LinearRegression()

// Train the model using the training sets
regr.fit(X, Y)
regr.coef_
regr.intercept_

import numpy as np
import statsmodels.api as sm

// Single Y
YY = Y[:,40000]
mod = sm.OLS(YY, X)
res = mod.fit()
print res.summary()

BenjaminHsieh commented:
ok I will use this to write test functions for regression

changsiyao committed:
update imports and getGainLoss

pigriver123 commented:
思瑶：

sorry，我不是很会传github,明天一定好好学学。我写好了那个check的function,同时regression full
也有小点点修改。我把两个文件先发给你。

On Thu, Nov 19, 2015 at 12:46 AM, changsiyao notifications@github.com
wrote:

> Also, here are some code if we try to use sklearn or statsmodels.api
> package.
> Didn't want to pull request on this.
> 
> from sklearn import datasets, linear_model
> X, Y are constructed like the way in calcBeta(data, gain, loss).
> 
> X = np.ones((len(gain), 3))
> X[:, 0] = gain
> X[:, 1] = loss
> designp = npl.pinv(design)
> 
> T = data.shape[-1]
> Y = np.reshape(data, (-1, T)).T
> Create linear regression object
> 
> regr = linear_model.LinearRegression()
> Train the model using the training sets
> 
> regr.fit(X, Y)
> regr.coef_
> regr.intercept_
> 
> import numpy as np
> import statsmodels.api as sm
> Single Y
> 
> YY = Y[:,40000]
> mod = sm.OLS(YY, X)
> res = mod.fit()
> print res.summary()
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/berkeley-stat159/project-theta/pull/50#issuecomment-157991143
> .

changsiyao committed:
fixed inputs in calcBeta

changsiyao committed:
fix bug

changsiyao committed:
fix bug
